docs/architecture/adr-030-msg-auth-module.md

```diff
@@ -0,0 +1,222 @@
+	// empty, there is no spend limit and any amount of coins can be spent.
+	SpendLimit sdk.Coins
+}
+
+func (cap SendAuthorization) MethodName() string {
+	return "/cosmos.bank.v1beta1.Msg/Send"
+}
+
+func (cap SendAuthorization) Accept(msg sdk.ServiceMsg, block abci.Header) (allow bool, updated Authorization, delete bool) {
+	switch req := msg.Request.(type) {
+	case bank.MsgSend:
+		left, invalid := cap.SpendLimit.SafeSub(req.Amount)
+		if invalid {
+			return false, nil, false
+		}
+		if left.IsZero() {
+			return true, nil, true
+		}
+		return true, SendAuthorization{SpendLimit: left}, false
+	}
+	return false, nil, false
+}
+```
+
+A different type of capability for `MsgSend` could be implemented
+using the `Authorization` interface with no need to change the underlying
+`bank` module.
+
+### `Msg` Service
+
+```proto
+service Msg {
+  // GrantAuthorization grants the provided authorization to the grantee on the granter's
+  // account with the provided expiration time.
+  rpc GrantAuthorization(MsgGrantAuthorization) returns (MsgGrantAuthorizationResponse);
+
+  // ExecAuthorized attempts to execute the provided messages using
+  // authorizations granted to the grantee. Each message should have only
+  // one signer corresponding to the granter of the authorization.
+  rpc ExecAuthorized(MsgExecAuthorized) returns (MsgExecAuthorizedResponse)
+
+
+  // RevokeAuthorization revokes any authorization corresponding to the provided method name on the
+  // granter's account that has been granted to the grantee.
+  rpc RevokeAuthorization(MsgRevokeAuthorization) returns (MsgRevokeAuthorizationResponse);
+}
+
+message MsgGrantAuthorization{
+  string granter = 1;
+  string grantee = 2;
+  google.protobuf.Any authorization = 3 [(cosmos_proto.accepts_interface) = "Authorization"];
+  google.protobuf.Timestamp expiration = 4;
+}
+
+message MsgExecAuthorized {
+    string grantee = 1;
+    repeated google.protobuf.Any msgs = 2;
+}
+
+message MsgRevokeAuthorization{
+  string granter = 1;
+  string grantee = 2;
+  string method_name = 3;
+}
+```
+
+### Router Middleware
+
+The `msg_authorization` `Keeper` will expose a `DispatchActions` method which allows other modules to send `ServiceMsg`s
+to the router based on `Authorization` grants:
+
+```go
+type Keeper interface {
+  DispatchActions(ctx sdk.Context, grantee sdk.AccAddress, msgs []sdk.ServiceMsg) sdk.Result`
+}
+```
+
+This allows the functionality provided by `msg_authorization` to be used for future inter-module object capabilities
+permissions as described in https://github.com/cosmos/cosmos-sdk.
+
+### CLI
+
+#### `--send-as` Flag
+
+When a CLI user wants to run a transaction on behalf of another account using `MsgExecAuthorized`, they
+can use the `--send-as` flag to specify the granter's account but signed by the grantee's account as
+specified by the `--from` flag. For instance `gaiacli tx gov vote 1 yes --from mykey --send-as cosmos3thsdgh983egh823`
```

Comment:
we cannot use  `gaiacli tx gov vote 1 yes --from mykey --send-as cosmos3thsdgh983egh823` to route the requests to respective handlers. We can do the following instead:
 `gaiacli tx gov vote 1 yes --from <grantee> --generate-only | gaiacli tx authz exec --send-as <granter> --from <grantee>`

where as authz exec command is, `tx authz exec <tx_json> --send-as <granter> --from <grantee>`

Reply:
Updated this text to use your example

Reply:
Thanks @aaronc , there's small change in the implementation. New CLI command would be, `gaiacli tx gov vote 1 yes --from <granter> --generate-only | gaiacli tx authz exec --from <grantee>`